Mathlib/Analysis/Complex/Hadamard.lean

```diff
@@ -0,0 +1,427 @@
+  apply Real.sSup_nonneg
+  rintro y ⟨z1, _, hz2⟩
+  simp only [← hz2, comp, norm_nonneg]
+
+/-- `sSup` of `norm` translated by `ε > 0` is positive applied to the image of `f` on the
+vertical line `re z = x` -/
+lemma sSupNormIm_eps_pos {ε : ℝ} (hε : ε > 0) (x : ℝ) : 0 < ε + sSupNormIm f x := by
+   linarith [sSupNormIm_nonneg f x]
+
+/-- Useful rewrite for the absolute value of `invInterpStrip`-/
+lemma abs_invInterpStrip {ε : ℝ} (hε : ε > 0) :
+    abs (invInterpStrip f z ε) =
+    (ε + sSupNormIm f 0) ^ (z.re - 1) * (ε + sSupNormIm f 1) ^ (-z.re) := by
+  simp only [invInterpStrip, map_mul]
+  repeat rw [← ofReal_add]
+  repeat rw [abs_cpow_eq_rpow_re_of_pos (sSupNormIm_eps_pos f hε _) _]
+  simp only [sub_re, one_re, neg_re]
+
+/-- The function `invInterpStrip` is `diffContOnCl`. -/
+lemma diffContOnCl_invInterpStrip {ε : ℝ} (hε : ε > 0) :
+    DiffContOnCl ℂ (fun z ↦ invInterpStrip f z ε) (verticalStrip 0 1) := by
+  apply Differentiable.diffContOnCl
+  apply Differentiable.mul
+  · apply Differentiable.const_cpow (Differentiable.sub_const (differentiable_id') 1) _
+    left
+    rw [← ofReal_add, ofReal_ne_zero]
+    simp only [ne_eq, ne_of_gt (sSupNormIm_eps_pos f hε 0), not_false_eq_true]
+  · apply Differentiable.const_cpow (Differentiable.neg differentiable_id')
+    apply Or.inl
+    rw [← ofReal_add, ofReal_ne_zero];
+    exact (ne_of_gt (sSupNormIm_eps_pos f hε 1))
+
+/-- If `f` is bounded on the unit vertical strip, then `f` is bounded by `sSupNormIm` there. -/
+lemma norm_le_sSupNormIm (f : ℂ → E) (z : ℂ) (hD : z ∈ verticalClosedStrip 0 1)
+    (hB : BddAbove ((norm ∘ f) '' (verticalClosedStrip 0 1))) :
+    ‖f z‖ ≤ sSupNormIm f (z.re) := by
+  refine le_csSup ?_ ?_
+  · apply BddAbove.mono (image_subset (norm ∘ f) _) hB
+    exact preimage_mono (singleton_subset_iff.mpr hD)
+  · apply mem_image_of_mem (norm ∘ f)
+    simp only [mem_preimage, mem_singleton]
+
+/-- The function `f` is bounded by `sSupNormIm`. (`f` is bounded on its domain.) -/
+lemma norm_lt_sSupNormIm_eps (f : ℂ → E) (ε : ℝ) (hε : ε > 0) (z : ℂ)
+    (hD : z ∈ verticalClosedStrip 0 1) (hB : BddAbove ((norm ∘ f) '' (verticalClosedStrip 0 1))) :
+    ‖f z‖ < ε + sSupNormIm f (z.re) :=
+  lt_add_of_pos_of_le hε (norm_le_sSupNormIm f z hD hB)
+
+/-- When the function `f` is bounded above on a vertical strip, then so is `F`. -/
+lemma F_BddAbove (f : ℂ → E) (ε : ℝ) (hε : ε > 0)
+    (hB : BddAbove ((norm ∘ f) '' (verticalClosedStrip 0 1))) :
+    BddAbove ((norm ∘ (F f ε)) '' (verticalClosedStrip 0 1)) := by
+ -- Rewriting goal
+  simp only [F, image_congr, comp_apply, map_mul, invInterpStrip]
+  rw [bddAbove_def] at *
+  cases' hB with B hB
+  -- Using bound
+  use ((max 1 ((ε + sSupNormIm f 0) ^ (-(1 : ℝ)))) * max 1 ((ε + sSupNormIm f 1) ^ (-(1 : ℝ)))) * B
+  simp only [mem_image, forall_exists_index, and_imp, forall_apply_eq_imp_iff₂]
+  intros z hset
+  specialize hB (‖f z‖)
+  specialize hB _
+  · simp only [image_congr, mem_image, comp_apply]; use z
```

Comment:
It's weird to specialize twice in a row like that. Why not just do it in one go, with
```
  specialize hB (‖f z‖) (by simpa [image_congr, mem_image, comp_apply] using ⟨z, hset, rfl⟩)
```
Also this means you just have one goal, so you don't need to indent more.